### PR TITLE
[python] Add and improve documentation for many public-facing methods

### PR DESCRIPTION
**Pull request by thetorpedodog**
- Adds many missing docstrings
- Reformats unwrapped docstrings to wrap around 80–88 lines for easier onscreen reading
- Converts some docstrings to be (closer to?) Sphinx format.

Part of #858 and #602.

### CODE REVIEW

File: apis/python/src/tiledbsoma/collection.py [hunk 1]
```diff
@@ -76,6 +76,17 @@ def create(
...
     ) -> _Self:
+        """Creates and opens a new SOMA collection in storage.
+
+        This creates a new SOMA collection of the current type in storage and
+        returns it opened for writing.
+
+        :param uri: The location to create this SOMA collection at.
+        :param platform_config: Optional call-specific options to use when
+            creating this collection. (Currently unused.)
+        :param context: If provided, the Context to use when creating and
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/collection.py:87]: the Context -> the `SOMATileDBContext`

File: apis/python/src/tiledbsoma/collection.py [hunk 2]
```diff
@@ -190,6 +225,13 @@ def _add_new_ndarray(
         shape: Sequence[int],
         platform_config: Optional[options.PlatformConfig] = None,
     ) -> _NDArr:
+        """Adds a new NDArray to this Collection.
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/collection.py:228]: Adds a new `SparseNDArray` or `DenseNDArray`

(These help-strings are user-facing and there is no user-facing `NDArray` class)

File: apis/python/src/tiledbsoma/dataframe.py [hunk 3]
```diff
@@ -24,7 +24,9 @@ class DataFrame(TileDBArray, somacore.DataFrame):
     """
     Represents ``obs``, ``var``, and others.
 
-    All ``DataFrame`` must contain a column called ``soma_joinid``, of type ``int64``. The ``soma_joinid`` column contains a unique value for each row in the ``DataFrame``, and intended to act as a joint key for other objects, such as ``SparseNDArray``.
+    All ``DataFrame`` must contain a column called ``soma_joinid``, of type ``int64``.
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/dataframe.py:27]: All -> Every

File: apis/python/src/tiledbsoma/dataframe.py [hunk 4]
```diff
@@ -44,11 +46,18 @@ def create(
 
         [lifecycle: experimental]
 
-        :param schema: Arrow Schema defining the per-column schema. This schema must define all columns, including columns to be named as index columns. If the schema includes types unsupported by the SOMA implementation, an error will be raised.
+        :param schema: Arrow Schema defining the per-column schema. This schema
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/dataframe.py:49]: Arrow Schema -> Arrow schema

File: apis/python/src/tiledbsoma/dataframe.py [hunk 5]
```diff
@@ -106,19 +115,21 @@ def read(
         platform_config: Optional[options.PlatformConfig] = None,
     ) -> TableReadIter:
         """
-        Read a user-defined subset of data, addressed by the dataframe indexing columns, optionally filtered, and return results as one or more Arrow.Table.
+        Read a user-defined subset of data, addressed by the dataframe indexing columns,
+        optionally filtered, and return results as one or more Arrow.Table.
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/dataframe.py:119]: Arrow.Table -> `Arrow.Table`

File: apis/python/src/tiledbsoma/dataframe.py [hunk 6]
```diff
@@ -213,11 +224,15 @@ def write(
         self, values: pa.Table, platform_config: Optional[Mapping[str, Any]] = None
     ) -> None:
         """
-        Write an Arrow.Table to the persistent object. As duplicate index values are not allowed, index values already present in the object are overwritten and new index values are added.
+        Write an Arrow.Table to the persistent object. As duplicate index values
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/dataframe.py:227]: Arrow.Table -> `Arrow.Table`

Comment (johnkerl) [reply apis/python/src/tiledbsoma/dataframe.py:227]: and below; I won't flag every instance

File: apis/python/src/tiledbsoma/dataframe.py [hunk 7]
```diff
@@ -243,10 +258,10 @@ def write(
 def _canonicalize_schema(
     schema: pa.Schema, index_column_names: Sequence[str]
 ) -> pa.Schema:
-    """
-    Handle default column additions (e.g., soma_joinid) and error checking on required columns.
+    """Turn a pyarrow Schema into the canonical version and check for errors.
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/dataframe.py:261]: pyarrow Schema -> `pyarrow.Schema`

File: apis/python/src/tiledbsoma/dataframe.py [hunk 8]
```diff
@@ -307,6 +322,7 @@ def _build_tiledb_schema(
     tiledb_create_options: TileDBCreateOptions,
     context: SOMATileDBContext,
 ) -> tiledb.ArraySchema:
+    """Converts a pyarrow.Schema into a TileDB ArraySchema for creation."""
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/dataframe.py:325]: ditto

File: apis/python/src/tiledbsoma/dense_nd_array.py [hunk 9]
```diff
@@ -20,25 +19,6 @@ class DenseNDArray(NDArray, somacore.DenseNDArray):
     [lifecycle: experimental]
     """
 
-    @property
-    def shape(self) -> NTuple:
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/dense_nd_array.py:24]: whaaaat? at a glance it looks like these went missing ... i'm trying to see where they went to on this PR ... 👀

File: apis/python/src/tiledbsoma/query_condition.py [hunk 10]
```diff
@@ -91,20 +91,20 @@ class QueryCondition:
...
-    >>>     # Note precedence is equivalent to:
-    >>>     # tiledbsoma.QueryCondition("foo > 5 or ('asdf' == attr('b a r') and baz <= val(1.0))")
-    >>>     A.query(cond="foo > 5 or 'asdf' == attr('b a r') and baz <= val(1.0)")
-    >>>
-    >>>     # Select cells where the attribute values for `foo` are equal to
-    >>>     # 1, 2, or 3.
-    >>>     # Note this is equivalent to:
-    >>>     # tiledbsoma.QueryCondition("foo == 1 or foo == 2 or foo == 3")
-    >>>     A.query(cond="foo in [1, 2, 3]")
+    Example::
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/query_condition.py:94]: double colon? TIL! :)

File: apis/python/src/tiledbsoma/sparse_nd_array.py [hunk 11]
```diff
@@ -151,7 +149,7 @@ def write(
 
         [lifecycle: experimental]
 
-        Arrow sparse tensor: the coordinates in the Arrow SparseTensor will be interpreted
+        Arrow sparse tensor: the coordinates in the Arrow SparseTensor are interpreted
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/sparse_nd_array.py:152]: 💯

**Review by johnkerl — CHANGES_REQUESTED** [review]

File: apis/python/src/tiledbsoma/dataframe.py (hunk 6 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/dataframe.py:227]: ibid.

File: apis/python/src/tiledbsoma/dataframe.py (hunk 7 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/dataframe.py:261]: "Arrow schema"

File: apis/python/src/tiledbsoma/dataframe.py (hunk 8 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/dataframe.py:325]: ibid.

File: apis/python/src/tiledbsoma/dense_nd_array.py (hunk 9 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/dense_nd_array.py:24]: These were exact duplicates of the methods they overrode; thus they do not need separate documentation after being removed.

File: apis/python/src/tiledbsoma/query_condition.py (hunk 10 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/query_condition.py:94]: It's a Sphinx directive to treat the next indented block like Python code with highlighting and everything.

File: apis/python/src/tiledbsoma/collection.py (hunk 2 above)
Comment (johnkerl) [reply apis/python/src/tiledbsoma/collection.py]: I disagree and maintain my position. However I won't block the PR for this.

Comment (bkmartinjr) [reply apis/python/src/tiledbsoma/collection.py]: I agree with Paul. IMHO, it is not confusing, and the brevity is appreciated.

File: apis/python/src/tiledbsoma/query_condition.py [hunk 12]
```diff
@@ -91,20 +91,20 @@ class QueryCondition:
...
+
+        with tiledb.open(uri, mode="r") as A:
```
Comment (bkmartinjr) [inline apis/python/src/tiledbsoma/query_condition.py:96]: It is quite confusing (to me!) to have a TileDB-Py example in the middle of the `tiledbsoma` package. It suggest that the TileDB-Py syntax/classes are workable, eg., I can hand a QueryCondition instance to a `tiledbsoma` class, or something?

Options:
* [best] rewrite to use the SOMA API
* delete

**Review by bkmartinjr — CHANGES_REQUESTED** [review]
changes look good, except for my cognitive dissonance around the `TileDB-Py` API example in the `tiledbsoma` library.  That definitely needs to be aligned with the package API.

File: apis/python/src/tiledbsoma/query_condition.py (hunk 12 above)
Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/query_condition.py]: Fixed, I believe. Please double-check my syntax here.

Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/query_condition.py]: Fixed, I believe. Please double-check my syntax here.

Comment (thetorpedodog) [reply apis/python/src/tiledbsoma/query_condition.py]: love the good website that sometimes posts comments twice

**Comment by thetorpedodog** [comment]
Ideally I would like to get #902 in before this one; that will remove the somewhat out-of-place removal of the duplicate functions from whichever NDArray class they were removed from.

**Review by bkmartinjr — APPROVED** [review]

**Pull request merged**